extension/spark-sink/src/main/scala/org/apache/spark/sql/palo/PaloBulkLoadTask.scala

```diff
@@ -0,0 +1,276 @@
+import java.nio.ByteBuffer
+
+import org.json4s._
+import org.json4s.jackson.JsonMethods._
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.{AnalysisException, SparkSession}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.types.{StringType, StructType}
+
+/**
+ * Use `BULK LOAD` to load data to Palo by using http protocol to send data.
+ * To use http `PUT` data to palo, we need write intermediate contents to tmp files on
+ * local file system and when load finished, delete them
+ *
+ * Note: the size of tmp file should not exceed 1GB (advise)
+ */
+private[palo] class PaloBulkLoadTask (
+    sparkSession: SparkSession,
+    batchId: Long,
+    checkpointRoot: String,
+    parameters: Map[String, String],
+    schema: StructType)
+  extends PaloWriteTask(sparkSession, batchId, checkpointRoot, parameters, schema) with Logging {
+
+  // use the "label" as the file name which will send to palo by http protocol
+  // the tmp file to store data generated by Dataset
+  private[palo] var file = new File(label)
+  private[palo] val httpMaxRetries: Int =
+      parameters.getOrElse(PaloConfig.HTTP_MAX_RETRIES, "3").toInt
+  private[palo] val httpConnectTimeoutMs: Int =
+      parameters.getOrElse(PaloConfig.HTTP_CONNECT_TIMEOUT_MS, "60000").toInt
+  private[palo] val httpReadTimeoutMs: Int =
+      parameters.getOrElse(PaloConfig.HTTP_READ_TIMEOUT_MS, "10000").toInt
+  private[palo] val bulkLoadReadBufferSize: Int =
+      parameters.getOrElse(PaloConfig.BULK_LOAD_READ_BUFFERSIZE, "1048576").toInt // 1MB
+  private[palo] val httpPort: String = parameters.getOrElse(PaloConfig.HTTP_PORT, "8030")
+  private[palo] val separator: String =
+      parameters.getOrElse(PaloConfig.SEPARATOR, ",") // default ","
+  private[palo] val maxFilterRatio =
+      parameters.getOrElse(PaloConfig.MAX_FILTER_RATIO, "0").toDouble
+  assert(maxFilterRatio >= 0 && maxFilterRatio <= 1, "maxFilterRatio must between [0,1]")
+
+  private[palo] val urlSuffix = s"api/${database}/${table}/_load" +
+          s"?label=${label}&${PaloConfig.MAX_FILTER_RATIO}=${maxFilterRatio}" +
+          s"&column_separator=${separator}"
+
+  private[palo] var outputStream: FileOutputStream = null
+  private[palo] var writer: BufferedWriter = null
+
+  // init outputStream to write tmp file
+  init
+
+  private def init() {
+    try {
+      // dataPath may exists for failover, delete the dataPath and create new one firstly
+      if (file.exists) {
+        file.delete
+        logInfo(s"file:${file.getAbsolutePath} exists, delete it successfully")
+      }
+      assert(file.createNewFile, "create file failed when init")
+      outputStream = new FileOutputStream(file)
+      writer = new BufferedWriter(new OutputStreamWriter(outputStream, "UTF-8"))
+
+      logInfo(s"init successfully, create a new file:${file.getAbsolutePath} to store contents")
+    } catch {
+      case e: IOException =>
```

Comment:
maybe should add some handler such as LOG, user-defined handler ，ABORT

Reply:
sorry , I don't get your idea here, Orz